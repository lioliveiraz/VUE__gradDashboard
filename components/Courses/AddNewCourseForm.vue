<template>
  <div class="relative w-auto my-6 mx-auto h-auto">
    <div class="g-modal-form--inner focus:outline-none">
      <div class="border-blueGray-200 g-modal-form--header">
        <h3 data-testId="course_form-title">Update Path</h3>
      </div>
      <div class="g-modal-form--content">
        <form @submit="handleSubmit" data-testId="course_form" :key="key">
          <div>
            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.NUMBER_INPUT,
                name: this.TABLE_HEAD_WEEK_ENGLISH,
                placeholder: 'week number',
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.COURSE_CODE_INPUT,
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.COURSE_NAME_INPUT,
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.TABLE_HEAD_SOURCE_ENGLISH,
                placeholder: 'Udemy',
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.NUMBER_INPUT,
                name: this.TABLE_HEAD_DURATION_ENGLISH,
                placeholder: '875463',
                required: true,
              }"
            />
            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.CHECKBOX_INPUT,
                name: this.ASSESSMENT_INPUT,
              }"
            />

            <div class="g-addNewcourse-buttons">
              <input
                type="submit"
                value="Send"
                :class="
                  isFormValid ? 'g-base-btn-blue' : 'g-base-btn-blue g-disabled'
                "
                :disabled="!isFormValid"
              />
              <BaseButton :handleClick="toggleComponent" value="close" />
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</template>

<script>
import BaseInput from "../BaseInput";
import { addCourse } from "../../api/requests/post";
import { mapGetters, mapActions } from "vuex";
import { courseValidation } from "../../helpers/validation";
import BaseButton from "../../components/Style/BaseButton";
import { isObjectEmpty, isObjectValuesEmpty } from "../../helpers/service";

export default {
  components: { BaseInput, BaseButton },
  props: { toggleComponent: Function },

  data() {
    return {
      courseData: {
        assessment: false,
        link:
          "https://giphy.com/gifs/makespace-cat-yoga-xUPGcyi4YxcZp8dWZq/tile",
      },
      key: 0,
      isFormValid: false,
    };
  },
  computed: {
    ...mapGetters("auth", ["getToken"]),
  },
  methods: {
    ...mapActions("courses", ["fetchCourses", "handleAddCourse"]),

    formValidation() {
      const errors = courseValidation(this.courseData);
      this.isFormValid = isObjectEmpty(errors);
    },
    getUserInput(inputValue, inputName) {
      this.courseData[inputName] = inputValue;
      if (
        !isObjectEmpty(this.courseData) &&
        isObjectValuesEmpty(this.courseData).length === 7
      ) {
        this.formValidation();
      } else {
        this.isFormValid = false;
      }
    },
    async handleSubmit(e) {
      e.preventDefault();
      this.courseData[this.DURATION_INPUT] = +this.courseData[
        this.DURATION_INPUT
      ];
      if (!this.isFormValid) return null;
      addCourse(this.courseData, this.getToken)
        .then((res) => {
          this.errors = {};
          this.toggleComponent();
          this.key = this.key + 1;
          this.handleAddCourse(this.courseData);
          this.$toast(res.data.message, { type: this.TOAST_SUCCESS });
        })
        .catch((err) =>
          this.$toast(err.response.data.message, { type: this.TOAST_ERROR })
        );
    },
  },
};
</script>

<style scoped>
</style>