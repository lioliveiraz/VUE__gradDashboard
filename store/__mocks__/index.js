import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export const getters = {
    isLoggedIn: jest.fn().mockReturnValue(true),
    getToken: jest.fn().mockReturnValue({ token: "token" }),
    getName: jest.fn().mockReturnValue(),
    getCourses: jest.fn().mockReturnValue([{ name: "course" }]),
    getAssessments: jest.fn().mockReturnValue([{ name: "assessments" }]),
    getScores: jest.fn().mockReturnValue([{ name: "scores" }])

};

export const mutations = {
    SET_TOKEN: jest.fn(),
    SET_ID: jest.fn(),
    SET_NAME: jest.fn(),
    SET_COURSES: jest.fn(),
    SET_ASSESSMENTS: jest.fn(),
    SET_SCORES: jest.fn(),
    ADD_COURSE: jest.fn(),
    ADD_ASSESSMENT: jest.fn(),
    ADD_SCORE: jest.fn()
};

export const actions = {
    login: jest.fn(),
    logout: jest.fn(),
    fetchCourses: jest.fn(),
    fetchScores: jest.fn(),
    handleAddCourse: jest.fn(),
    handleAddScore: jest.fn()


};

export const state = {
    token: "token",
    user_id: 1,
    user_name: "name",
    courses: [{ name: "course" }],
    assessments: [{ name: "assessments" }],
    scores: [{ name: "scores" }]
};

export function __createMocks(custom = { getters: {}, mutations: {}, actions: {}, state: {} }) {
    const mockGetters = Object.assign({}, getters, custom.getters);
    const mockMutations = Object.assign({}, mutations, custom.mutations);
    const mockActions = Object.assign({}, actions, custom.actions);
    const mockState = Object.assign({}, state, custom.state);

    return {
        getters: mockGetters,
        mutations: mockMutations,
        actions: mockActions,
        state: mockState,
        store: new Vuex.Store({
            modules: {
                auth: {
                    namespaced: true,
                    getters: mockGetters,
                    mutations: mockMutations,
                    actions: mockActions,
                    state: mockState,
                },
                courses: {
                    namespaced: true,
                    getters: mockGetters,
                    mutations: mockMutations,
                    actions: mockActions,
                    state: mockState,
                }
            }

        }),
    };
}

export const store = __createMocks().store;