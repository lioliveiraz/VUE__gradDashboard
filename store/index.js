import Vuex from 'vuex';
import Vue from 'vue';
import auth from './modules/auth';

Vue.use(Vuex);


const store = () => {
    return new Vuex.Store({
        modules: { auth }
    });
};

export default store;