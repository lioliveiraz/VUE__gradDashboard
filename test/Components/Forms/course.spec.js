import Vuex from 'vuex';
import Vue from 'vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { __createMocks as createStoreMocks, store } from '../../../store/__mocks__';
import AddNewCourseForm from '../../../components/Courses/AddNewCourseForm.vue';


jest.mock('../../../store/__mocks__');

const localVue = createLocalVue();

localVue.use(Vuex);

describe('<AddNewCourseForm/>', () => {
    let wrapper;
    beforeEach(async () => {
        wrapper = await shallowMount(AddNewCourseForm, {
            store: store,
            localVue
        });
    });

    it('should render correctly', () => {
        const idArr = ["course_form", "course_form-title"];
        idArr.forEach((id) => expect(wrapper.find(id)).toBeTruthy());
        expect(wrapper).toMatchSnapshot();
    });

});