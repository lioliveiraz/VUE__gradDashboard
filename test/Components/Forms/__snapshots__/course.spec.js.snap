// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddNewCourseForm/> should render correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
